api/v1alpha1/openstackimage_types.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// See https://docs.openstack.org/glance/latest/admin/useful-image-properties.html
+// for a list of 'well known' image properties we might consider supporting explicitly.
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=255
+type ImageTag string
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=10
+type ImageContainerFormat string
+
+// +kubebuilder:validation:MinLength:=1
+// +kubebuilder:validation:MaxLength:=10
+type ImageDiskFormat string
+
+// kubebuilder:validation:Enum:=public;private;shared;community
```

Comment:
Ack. That would presumably also apply to the hash algorithms.

Reply:
FYI I initially made this change, then after I'd lived with it for a while I reverted it. I did this because I found myself having to write conversions for every one of these in code, and realising that users will also mentally have to do the same thing. This is a proxy API, so these strings only have meaning in OpenStack glance, which is the documentation a user is most likely to be looking at when working out what to put here. If I Pascal-ise them, they also need to mentally translate the glance value to whatever I picked here.

I think it's a case of 'pick what you're going to be inconsistent with'. I think following the API we're proxying is less bad here.